libtiff/tests/get_tag_defaults.py

```diff
@@ -0,0 +1,112 @@
+            (lt.TIFFTAG_IMAGEWIDTH, 'ImageWidth', 128),
+        ]
+    },
+    'float': {
+        'ctype': ctypes.c_float,
+        'set_wrapper': ctypes.c_double,
+        'values': [
+            (lt.TIFFTAG_XRESOLUTION, 'XResolution', 88.0),
+            (lt.TIFFTAG_YRESOLUTION, 'YResolution', 88.0),
+            (lt.TIFFTAG_XPOSITION, 'XPosition', 88.0),
+            (lt.TIFFTAG_YPOSITION, 'YPosition', 88.0),
+        ]
+    },
+    'double': {
+        'ctype': ctypes.c_double,
+        'set_wrapper': ctypes.c_double,
+        'values': [
+            (lt.TIFFTAG_SMAXSAMPLEVALUE, 'SMaxSampleValue', 255.0),
+            (lt.TIFFTAG_SMINSAMPLEVALUE, 'SMinSampleValue', 0.0),
+        ]
+    },
+    'string': {
+        'ctype': ctypes.c_char_p,
+        'values': [
+            (lt.TIFFTAG_ARTIST, 'Artist', b"test string"),
+            (lt.TIFFTAG_DATETIME, 'DateTime', b"test string"),
+            (lt.TIFFTAG_HOSTCOMPUTER, 'HostComputer', b"test string"),
+            (lt.TIFFTAG_IMAGEDESCRIPTION, 'ImageDescription', b"test string"),
+            (lt.TIFFTAG_MAKE, 'Make', b"test string"),
+            (lt.TIFFTAG_MODEL, 'Model', b"test string"),
+            (lt.TIFFTAG_SOFTWARE, 'Software', b"test string"),
+        ]
+    }
+}
+
+
+def get_default_tag_values(tmp_path):
+    ltc = lt.libtiff
+    tiff = lt.TIFF.open(tmp_path / 'default_values.tiff', mode='w')
+    defaults = {}
+
+    data_holders = {
+        'uint16': ctypes.c_uint16(0),
+        'uint32': ctypes.c_uint32(0),
+        'float': ctypes.c_float(0.0),
+        'double': ctypes.c_double(0.0),
+        'string': ctypes.c_char_p(b''),
+    }
+
+    for type_name, test_data in tag_test_data.items():
+        for tag_const, _, _ in test_data['values']:
+            if tag_const in defaults:
+                continue  # Already checked
+
+            data_holder = data_holders[type_name]
+            p_data_holder = ctypes.byref(data_holder)
+
+            if ltc.TIFFGetFieldDefaulted(tiff, tag_const, p_data_holder):
+                defaults[tag_const] = data_holder.value
+            else:
+                defaults[tag_const] = None
+    tiff.close()
+    # manually add missing default
+    defaults[lt.TIFFTAG_SAMPLEFORMAT] = 1
+    return defaults
+
+
+with tempfile.TemporaryDirectory() as tmpdir:
+    default_values = get_default_tag_values(pathlib.Path(tmpdir))
+
+    new_tag_test_data = {}
+    for type_name, test_data in tag_test_data.items():
+        new_tag_test_data[type_name] = {
+            'ctype': test_data['ctype'],
+            'values': []
+        }
+        if 'set_wrapper' in test_data:
+            new_tag_test_data[type_name]['set_wrapper'] = test_data['set_wrapper']
+
+        for tag_const, tag_name, value in test_data['values']:
+            new_tag_test_data[type_name]['values'].append(
+                (tag_const, tag_name, value, default_values.get(tag_const))
+            )
+
+    print("tag_test_data = " + pprint.pformat(new_tag_test_data))
```

Comment:
So this script is only run once (which you've done already) to generate the defaults which you then pasted into the test module, right?

If defaults change between libtiff versions, wouldn't we want this to run every time the tests are run? In which case, could this be put this into a pytest fixture with session scope maybe? It could then use `tmp_path_factory` instead of `TemporaryDirectory`.

Reply:
Done